packages/react-dom/src/client/ReactDOM.js

```diff
@@ -667,14 +667,36 @@ const DOMRenderer = ReactFiberReconciler({
   },
 
   shouldSetTextContent(type: string, props: Props): boolean {
-    return (
-      type === 'textarea' ||
+    if (type === 'textarea') {
+      return true;
+    }
+
+    if (
       typeof props.children === 'string' ||
-      typeof props.children === 'number' ||
-      (typeof props.dangerouslySetInnerHTML === 'object' &&
-        props.dangerouslySetInnerHTML !== null &&
-        typeof props.dangerouslySetInnerHTML.__html === 'string')
-    );
+      typeof props.children === 'number'
+    ) {
+      return true;
+    }
+
+    if (
+      typeof props.dangerouslySetInnerHTML === 'object' &&
+      props.dangerouslySetInnerHTML !== null
+    ) {
+      if (typeof props.dangerouslySetInnerHTML.__html === 'string') {
+        return true;
+      }
+
+      if (props.dangerouslySetInnerHTML.__html === null) {
+        return false;
+      } else if (
+        typeof props.dangerouslySetInnerHTML.__html === 'object' &&
+        typeof props.dangerouslySetInnerHTML.__html.toString === 'function'
```

Comment:
I don't understand how this could work. Any JS object has toString defined.

Reply:
`Object.create(null).toString` maybe ?

Reply:
@gaearon good catch, could refactor to use `.hasOwnProperty`

Reply:
It could be inherited from something else. I think in general trying to detect a custom `toString` is a flawed approach.

Reply:
Hmm yeah, so maybe back to square 1? Just check for an object, with a check for the existence of `toString` for corner cases like `Object.create(null)`?